File: src/main/java/mnm/mods/tabbychat/client/gui/component/GuiLabel.java
package mnm.mods.tabbychat.client.gui.component;

import com.mojang.blaze3d.platform.GlStateManager;
import mnm.mods.tabbychat.util.ILocation;
import mnm.mods.tabbychat.util.text.FancyFontRenderer;
import net.minecraft.client.Minecraft;
import net.minecraft.util.text.ITextComponent;

/**
 * Gui component label used to show text on the screen.
 *
 * @author Matthew
 */
public class GuiLabel extends GuiComponent {

    private FancyFontRenderer fr;
    private ITextComponent text;
    private float angle;

    public GuiLabel() {
        this.fr = new FancyFontRenderer(Minecraft.getInstance().fontRenderer);
    }

    /**
     * Creates a label from a chat component.
     *
     * @param chat The text
     */
    public GuiLabel(ITextComponent chat) {
        this();
        this.setText(chat);
    }

    @Override
    public void render(int mouseX, int mouseY, float parTicks) {
        if (getText() == null)
            return;
        GlStateManager.pushMatrix();

        GlStateManager.rotatef(angle, 0, 0, angle);
        if (angle < 180) {
            GlStateManager.translated(-angle / 1.5, -angle / 4, 0);
        } else {
            GlStateManager.translated(-angle / 15, angle / 40, 0);
        }

        ILocation loc = getLocation();
        fr.drawChat(getText(), loc.getXPos() + 3, loc.getYPos() + 3, getPrimaryColorProperty().getHex(), true);

        GlStateManager.popMatrix();
    }

    /**
     * Sets the string of this label
     *
     * @param text The string
     */
    public void setText(ITextComponent text) {
        this.text = text;
    }

    /**
     * Gets the string of this label
     *
     * @return The string
     */
    public ITextComponent getText() {
        return text;
    }

    public float getAngle() {
        return angle;
    }

    public void setAngle(float angle) {
        this.angle = angle;
    }

}

File: src/main/kotlin/mnm/mods/tabbychat/client/gui/component/GuiLabel.kt
package mnm.mods.tabbychat.client.gui.component

import com.mojang.blaze3d.systems.RenderSystem
import mnm.mods.tabbychat.util.mc
import mnm.mods.tabbychat.util.text.FancyFontRenderer
import net.minecraft.util.text.ITextComponent

/**
 * Gui component label used to show text on the screen.
 *
 * @author Matthew
 */
open class GuiLabel(var text: ITextComponent? = null) : GuiComponent() {

    private val fr: FancyFontRenderer = FancyFontRenderer(mc.fontRenderer)

    var angle: Float = 0.toFloat()

    override fun render(x: Int, y: Int, parTicks: Float) {
        if (text == null)
            return
        RenderSystem.pushMatrix()

        RenderSystem.rotatef(angle, 0f, 0f, angle)
        if (angle < 180) {
            RenderSystem.translated(-angle / 1.5, (-angle / 4).toDouble(), 0.0)
        } else {
            RenderSystem.translated((-angle / 15).toDouble(), (angle / 40).toDouble(), 0.0)
        }

        val loc = location
        fr.drawChat(text!!, (loc.xPos + 3).toFloat(), (loc.yPos + 3).toFloat(), primaryColorProperty.hex, true)

        RenderSystem.popMatrix()
    }

}
